feat: Add gpcc::container::FixCapFIFO

DanielJerolm opened the pull request.

DanielJerolm committed:
Add implementation

DanielJerolm committed:
Add unittests

DanielJerolm committed:
Update history

Andreas143 commented on include/gpcc/container/FixCapFIFO.hpp:
```diff
@@ -0,0 +1,121 @@
+{
+  public:
+    FixCapFIFO(void) = delete;
+    FixCapFIFO(size_t const capacity);
```
Use SIZET instead of size_t to avoid imlicit type casts.

DanielJerolm replied:
`size_t` is the standard type to express the quantity of countable things like objects, bytes, etc.
I would like continue using it in the public API.

The user does not see any implicit cast, because `SIZET` is only used inside the implementation of `FixCapFIFO`. 

I would suggest that I make any implicit cast from `size_t` to `SIZET` explicit within the FIFO's implementation.

Andreas143 replied:
1) To make the implicit casts explicit is not my point. I do not prefer this.
2) If you want to use 'size_t' consistently in public API then this is ok.
3) My point is/was that a user constructs a FIFO object with a given capacity. The type of this 'capacity' parameter used by user could/should match to the concrete FIFO implementation. If the user asks the FIFO object for its capacity value (refer Capacity() method) then the returned type could/should be the same obviously. The same applies to Size() method.

If you choose solution 2) that's perfectly fine with me.

DanielJerolm replied:
Its 2)
All public methods use `size_t` to express capacity or size.

Andreas143 commented on include/gpcc/container/FixCapFIFO.hpp:
```diff
@@ -0,0 +1,121 @@
+    FixCapFIFO& operator=(FixCapFIFO const & rhv);
+    FixCapFIFO& operator=(FixCapFIFO && rhv) noexcept;
+
+    size_t Capacity(void) const noexcept;
+    size_t Size(void) const noexcept;
```
Use SIZET instead of size_t to avoid imlicit type casts.

DanielJerolm replied:
All public methods use size_t to express capacity or size.

Andreas143 commented on include/gpcc/container/FixCapFIFO.hpp:
```diff
@@ -0,0 +1,121 @@
+    T UnsafePop(void) noexcept;
+
+    size_t Push(T const * const pSrc, size_t const n) noexcept;
+    size_t Pop(T* const pDest, size_t const n) noexcept;
```
Use SIZET instead of size_t to avoid imlicit type casts.
Affects parameter type and return type.

DanielJerolm replied:
All public methods use size_t to express capacity or size.

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,563 @@
+/**
+ * \brief Move-assigns the content of another FIFO instance to this instance.
+ *
+ * \note  The capacity of the FIFO may be reduced or enlarged.
```
This description should be removed because the semantic of 'assignment' implies it.
Furthermore it leads to confusion because after assignment the destination shall be the same as the source before assignment. 
Note: The move assignment has the same semantics like copy assignment but it is 'optimized' for special use cases.
Note2: The behaviour (preconditions) of the copy and move assignment are different, but they shouldn't! Seems to be bad design.

DanielJerolm replied:
Note2: I agree, but the copy assignment operator may have to reallocate memory then.
Copy- and move-assignment operators are not meant to be used from interrupt context, so this no issue.
Maybe the CopyFromOther() function should be enhanced not to make a plain copy, but instead fill the target memory starting at the beginning.

DanielJerolm replied:
The note has been removed from the move-assignment operator.
The move assignment operator has no precondition.

The copy assignment operator has been revised. Its precondition has been weakened. Complete removal of the precondition would require a reallocation to increase the FIFO's capacity in some scenarios, which feels wrong for a _Fixed-Capacity_-FIFO.

What do you think?

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,563 @@
+ *
+ * \param rhv
+ * Unmodifiable reference to the other FIFO instance.\n
+ * Both FIFOs must have the same capacity.
```
This should be a precontition.

DanielJerolm replied:
The copy-assignment operator will be revised. The FIFOs do not need to have the same capacity, but the capacity of the destination FIFO must be sufficient for the current content of the source FIFO.

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,563 @@
+void FixCapFIFO<T, SIZET>::UnsafePush(T const value) noexcept
+{
+  spMemory_[wrIndex_] = value;
+  wrIndex_ = (wrIndex_ + 1U) & (~capacity_);
```
I suggest an alternative implementation for this line of code because it has (IMHO) some advantages. But lets discuss this point. Together with some other changes the limitation of capacity needs to be power of 2 could be removed by accepting a very smal performance penalty.
```suggestion
  if (wrIndex_ == (capacity_ - 1U)
    wrIndex_ = 0U;
  else
    wrIndex_++;
```

DanielJerolm replied:
The power-of-2 constraint has been removed from the FIFO's capacity. There is a small penalty (2 more instructions incl. a compare instruction), but that's acceptable.

Andreas143 commented on include/gpcc/container/FixCapFIFO.hpp:
```diff
@@ -0,0 +1,121 @@
+    void UnsafePush(T const value) noexcept;
+    T UnsafePop(void) noexcept;
+
+    size_t Push(T const * const pSrc, size_t const n) noexcept;
+    size_t Pop(T* const pDest, size_t const n) noexcept;
```
These functions handle multiple elements. This behaviour should be expressed by their name.
```suggestion
    size_t PushMultiple(T const * const pSrc, size_t const n) noexcept;
    size_t PopMultiple(T* const pDest, size_t const n) noexcept;
```

DanielJerolm replied:
Will be renamed

Andreas143 commented on include/gpcc/container/FixCapFIFO.hpp:
```diff
@@ -0,0 +1,121 @@
+    void Clear(void) noexcept;
+
+    void UnsafePush(T const value) noexcept;
+    T UnsafePop(void) noexcept;
```
I believe that providing 'unsafe' function is a bad idea if you want to enforce robust code.
There is no performance advantage if the user needs a single Push/Pop. He must check the precondition anyway!
If the user needs multiple Push/Pop then he should be encuraged to use the provided function for that use case.
Therefore add checks to the implementation and remove 'Unsafe' from function names.

DanielJerolm replied:
I suggest that I'll remove "unsafe".
I agree that the user must fulfill the preconditions in any case.

However, these methods are intended to be usable from interrupt context. There is no way to throw and no portable way to panic. `UnsafePush()` could return a bool to indicate success or (graceful) failure. `UnsafePop()` would require an output parameter for the popped item.

There is a frequent use-case behind this incorporating an ISR that transfers received bytes from a hardware FIFO into a software FIFO. I'll check if it can be solved without the need to invoke `IsFull()` if `Push()` is safe.

DanielJerolm replied:
The methods have been renamed to Push() and Pop() and they handle a full/empty FIFO gracefully.

Andreas143 commented on include/gpcc/container/FixCapFIFO.hpp:
```diff
@@ -0,0 +1,121 @@
+ * This class implements a light-weight FIFO with fixed capacity. It is designed for usage from both thread and
+ * interrupt context by the following measures:
+ * - Except for its constructor, it does not allocate memory at runtime.
+ * - Methods do not throw by guarantee.
```
This is not true. Currently the assignment operator may throw!

DanielJerolm replied:
Revised.

Andreas143 commented on include/gpcc/container/FixCapFIFO.hpp:
```diff
@@ -0,0 +1,121 @@
+ * context.
+ *
+ * # Optimizations and limitations
+ * Especially single push/pop operations are heavily optimized for ARM and x64 CPUs. The number of generated
```
AFAIK there is no 'heavy optimization' (by using -Os). It depends on the choosen compiler if there is a difference of one machine instruction or no difference at all.

DanielJerolm replied:
Section has been removed from doxygen. Power-of-2 constraint has been removed.

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,563 @@
+ * - - -
+ *
+ * \param rhv
+ * Unmodifiable reference to the other FIFO instance.\n
```
```suggestion
 * The other FIFO instance.\n
```

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,563 @@
+ * - - -
+ *
+ * \param other
+ * Unmodifiable reference to the other FIFO instance that shall be copied.
```
```suggestion
 * The other FIFO instance that shall be copied.
```

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,563 @@
+ * - - -
+ *
+ * \param rhv
+ * Universal reference to the other FIFO instance whose content shall be move-assigned to this instance.\n
```
```suggestion
 * The other FIFO instance whose content shall be move-assigned to this instance.\n
```

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,563 @@
+{
+  size_--;
+  auto const prevRdIdx = rdIndex_;
+  rdIndex_ = (rdIndex_ + 1U) & (~capacity_);
```
```suggestion
  if (rdIndex_ == (capacity_ - 1U)
    rdIndex_ = 0U;
  else
    rdIndex_++;
```

DanielJerolm replied:
The power-of-2 constraint on the capacity has been removed.

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,563 @@
+ * - - -
+ *
+ * \param pOtherData
+ * Pointer to the memory of the other FIFO. Data is copied from this location into this FIFO's storage according to
```
```suggestion
 * Pointer to the memory of the other FIFO. Data is copied from that location into this FIFO's storage according to
```

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,563 @@
+      // not full, wrap: |xxx    xxx|
+      //                  <b>    <a>
+      SIZET const a = (capacity_ - rdIndex_);
+      if (a != 0U)
```
This condition evaluates always to true. The values of a can be [1..capacity_], but never null.

DanielJerolm replied:
That's true. Will be removed.

DanielJerolm committed:
Apply suggestions from code review

Co-authored-by: Andreas143 <113170585+Andreas143@users.noreply.github.com>

DanielJerolm committed:
Apply review suggestion: Push/Pop -> Push/PopMultiple

DanielJerolm committed:
Fix review finding: Remove test for always true condition

DanielJerolm committed:
Copy assignment does not require same capacity.

The capacity of the destination FIFO must not be the
same as the capacity of the source FIFO. Instead it must
be sufficient for the current content of the source FIFO.

DanielJerolm committed:
Move assignment: Remove redundant note

DanielJerolm committed:
Revise UnsafePush/Pop -> Push/Pop (safe)

DanielJerolm committed:
Revise class' doxygen.

- Optimizations are no longer mentioned
- revise measures for usage in interrupt context

DanielJerolm committed:
Remove power-of-2-constraint on capacity

DanielJerolm committed:
Revise unittests

DanielJerolm committed:
Merge branch 'develop' into feature/fifo_container

DanielJerolm committed:
Merge branch 'develop' into feature/fifo_container

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,593 @@
+  if (&rhv != this)
+  {
+    if (capacity_ < rhv.size_)
+      throw std::logic_error("Insufficient capacity");
```
Why do you throw instead of trying to allocate memory? If allocations fails then bad_alloc exception is thrown. There is no difference from users point of view. But if the allocation is successful then the user is happy. In any case the copy assignment must not be used in interrupt context. Therefore I see no (real) drawback in trying to allocate memory.

DanielJerolm replied:
> If allocations fails then bad_alloc exception is thrown. There is no difference from users point of view. 

Good point! Will be fixed.

DanielJerolm committed:
Copy-assignment reallocates if capacity is not sufficient.

Andreas143 commented on include/gpcc/container/FixCapFIFO.tcc:
```diff
@@ -0,0 +1,596 @@
+    {
+      T* const pNewStorage = new T[rhv.size_];
+
+      capacity_ = rhv.size_;
```
You want to "copy" the whole FIFO object not only its data.
```suggestion
      T* const pNewStorage = new T[rhv.capacity_];

      capacity_ = rhv.capacity_;
```

DanielJerolm replied:
What if the capacity of the target is smaller, but sufficient?
In that case a reallocation does not make sense and a sparse copy is perfect.

Andreas143 replied:
Good point, but in this case it is realy better to avoid allocation. I would say that this is a special case. On the other hand if you need to allocate memory then there is no penalty if you allocate 'capacity'. Think about the case if the source FIFO has been created to hold 10 elements but during copy it holds only one ore zero elements. I would expect that the copy of that FIFO is able to hold 10 elements too OR (special case) the capacity (maybe 5) of the destination is not changed.

DanielJerolm replied:
The current implemented behavior is also aligned with the behavior of the copy-assignment operator of an `std::vector` (https://stackoverflow.com/questions/2663170/stdvector-capacity-after-copying).

The results of a Google search on "Does copy assignment of std::vector increase the capacity" also show that it is not common to copy capacity.

Andreas143 replied:
The behaviour of std::vector is not fully equal to (fixed capacity) FIFO! ;-)
Following situation:
FIFO_A has a (fixed) capacity of 5
FIFO_B has a (fixed) capacity of 10
Scenario 1: FIFO_B holds 3 elements and is copy assigned to FIFO_A
Expectation: FIFO_A -> capacity is still 5 (size is 3 obviously)
Scenario 2: FIFO_B holds 7 elements and is copy assigned to FIFO_A
Expectation: FIFO_A -> capacity is now 10 (size is 7 obviously)

According to your implementation the FIFO_A would have a capacity of 7 (suprise!) and it would be full (hey the source FIFO_B was not full :-( ).

DanielJerolm replied:
Thanks for the great example. An `std::vector` does not really have a full-state, that makes the difference.
But that means that the FixCapFIFO's copy-assignment operator **must always** ensure that the capacity of the target is always equal to or greater than the source' capacity.

Andreas143 replied:
Yes, this is the strict point of view that forbid the mentioned 'special case' (scenario 1) above.
I think it would be more clear (and easier for the user) if the capacity of the destination is equal or greater than the source after copy assignment.

DanielJerolm committed:
Copy-assignment includes adaption of capacity
